classes/client/do_client.php

```diff
@@ -0,0 +1,128 @@
+<?php
+// This file is part of Moodle - http://moodle.org/
+//
+// Moodle is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// Moodle is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Moodle.  If not, see <http://www.gnu.org/licenses/>.
+
+/**
+ * DigitalOcean Spaces client.
+ *
+ * @package   tool_objectfs
+ * @author    Brian Yanosik <kisonay@gmail.com>
+ * @license   http://www.gnu.org/copyleft/gpl.html GNU GPL v3 or later
+ */
+
+namespace tool_objectfs\client;
+
+defined('MOODLE_INTERNAL') || die();
+
+$autoloader = $CFG->dirroot . '/local/aws/sdk/aws-autoloader.php';
+
+if (!file_exists($autoloader)) {
+
+    // Stub class with bare implementation for when the SDK prerequisite does not exist.
+    class do_client {
+        public function get_availability() {
+            return false;
+        }
+
+        public function register_stream_wrapper() {
+            return false;
+        }
+    }
+
+    return;
+}
+
+require_once($autoloader);
+
+use Aws\S3\S3Client;
+
+class do_client extends s3_client {
+
+    public function __construct($config) {
+        $this->bucket = $config->do_space;
+        $this->set_client($config);
+    }
+
+    public function set_client($config) {
+        $this->client = S3Client::factory(array(
+            'credentials' => array('key' => $config->do_key, 'secret' => $config->do_secret),
+            'region' => $config->do_region,
+            'endpoint' => 'https://' . $config->do_region . '.digitaloceanspaces.com',
+            'version' => AWS_API_VERSION
+        ));
+    }
+
+    public function define_do_spaces_check($mform, $config) {
+        global $OUTPUT;
+        $connection = false;
+
+        $client = new do_client($config);
+
+        $connection = $client->test_connection();
+
+        if ($connection->success) {
+            $mform->addElement('html', $OUTPUT->notification($connection->message, 'notifysuccess'));
+
+            // Check permissions if we can connect.
+            $permissions = $client->test_permissions();
```

Comment:
Its my understanding you can't disable 'delete' functionality with DO spaces, so i would remove the delete check for DO